agent/metrics/monitor/prometheus_monitor.go

```diff
@@ -0,0 +1,126 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package monitor
+
+import (
+	"context"
+	"github.com/aws/amazon-ecs-agent/agent/acs/update_handler/os"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi"
+	"github.com/cihub/seelog"
+	dockercontainer "github.com/docker/docker/api/types/container"
+	"time"
+)
+
+const (
+	PrometheusContainerName = "Prometheus_Monitor"
+	dockerTimeout           = 2 * time.Minute
+)
+
+// Performs the required set-up for the Prometheus container. This function is
+// only called if PrometheusMetrics is enabled in the cfg. We perform the
+// following steps:
+// 1) Create Volume if not created.
+// 2) Create Prometheus container if not created.
+// 3) Start Prometheus container
+func InitPrometheusContainer(ctx context.Context, cfg *config.Config, dg dockerapi.DockerClient, fs ...os.FileSystem) string {
+	if !cfg.PrometheusMetricsEnabled {
+		return ""
+	}
+	prometheusVolumeSetup(dg, cfg, ctx)
+
+	containerID := cfg.PrometheusMonitorContainerID
+
+	// We should create the container in 2 situations: if containerID in state file
+	// was never initialized (Prometheus container never started) or containerID
+	// cannot be inspected. Note the fallthrough keyword here.
+	switch {
+	case containerID != "":
+		_, err := dg.InspectContainer(ctx, containerID, dockerTimeout)
+		if err == nil {
+			seelog.Infof("Prometheus container found with Docker ID: %s", containerID)
+			break
+		}
+		fallthrough
+	case containerID == "":
+		seelog.Info("Prometheus container not found. Creating one.")
+		// This is so we can mock the FileSystem and optionally pass it into this function
+		// for unit testing
+		var fsToUse os.FileSystem
+		if len(fs) > 0 {
+			fsToUse = fs[0]
+		} else {
+			fsToUse = os.Default
+		}
+		metadata, err := createPrometheusContainer(dg, cfg, ctx, fsToUse)
+		if err != nil {
+			seelog.Errorf("Prometheus container could not be created: %s", err.Error())
+			return ""
+		}
+		containerID = metadata.DockerID
+	}
+	// We start the Prometheus container after all set-up has been done
+	metadata := dg.StartContainer(ctx, containerID, dockerTimeout)
+	if metadata.Error == nil {
+		seelog.Info("Prometheus container started")
+	} else {
+		seelog.Errorf("Failed to start Prometheus container: %s", metadata.Error.Error())
+		return ""
+	}
+	return containerID
+}
+
+func prometheusVolumeSetup(dg dockerapi.DockerClient, cfg *config.Config, ctx context.Context) {
+	resp := dg.InspectVolume(ctx, cfg.PrometheusMonitorVolumeName, dockerTimeout)
+	if resp.Error != nil {
+		// Inspection failed, attempt to remove the volume.
+		dg.RemoveVolume(ctx, cfg.PrometheusMonitorVolumeName, dockerTimeout)
+		// Create the new volume
+		dg.CreateVolume(ctx, cfg.PrometheusMonitorVolumeName, "", nil, make(map[string]string), dockerTimeout)
```

Comment:
nit: i think it's probably better to check the error returned here and log something for both success and failure cases

Reply:
An error here is expected. I logged it as an info without using a failure-like message to avoid confusion on if the error is problematic.